src/zkp/pienc.rs

```diff
@@ -279,27 +275,204 @@ mod tests {
 
         let proof = PiEncProof::prove(rng, &input, &PiEncSecret { plaintext, nonce })?;
 
+        Ok((proof, input))
+    }
+
+    #[test]
+    fn proof_serializes_correctly() -> Result<()> {
+        let mut rng = crate::utils::get_test_rng();
+
+        let plaintext = random_plusminus_by_size(&mut rng, ELL);
+        let (proof, input) = build_random_proof(&mut rng, plaintext)?;
+
         let proof_bytes = bincode::serialize(&proof).unwrap();
         let roundtrip_proof: PiEncProof = bincode::deserialize(&proof_bytes).unwrap();
         let roundtrip_proof_bytes = bincode::serialize(&roundtrip_proof).unwrap();
+
         assert_eq!(proof_bytes, roundtrip_proof_bytes);
+        assert!(proof.verify(&input).is_ok());
+        assert!(roundtrip_proof.verify(&input).is_ok());
 
-        proof.verify(&input)
+        Ok(())
     }
 
     #[test]
-    fn test_paillier_encryption_in_range_proof() -> Result<()> {
+    fn plaintext_must_be_in_range() -> Result<()> {
         let mut rng = crate::utils::get_test_rng();
 
+        // A plaintext in the range 2^ELL should always succeed
         let in_range = random_plusminus_by_size(&mut rng, ELL);
+        let (proof, input) = build_random_proof(&mut rng, in_range)?;
+        assert!(proof.verify(&input).is_ok());
+
+        // A plaintext in range for encryption but larger (absolute value) than 2^ELL should fail
         let too_large =
             random_plusminus_by_size_with_minimum(&mut rng, ELL + EPSILON + 1, ELL + EPSILON)?;
+        let (too_large_proof, input) = build_random_proof(&mut rng, too_large.clone())?;
```

Comment:
I tried this and got interesting behavior. The proof expects an input in the range `±2^ell`, and it can guarantee that its in the range `±2^ell + epsilon`. When I tested it, I found mixed results for things in the range. I'm guessing it's kind of randomized for stuff in between -- like, it depends on the size of the challenge and the mask.

| input | result| 
| --- | --- |
| `2^ell` |  passes |
| `2^ell + 1` | passes |
| `2^(ell + epsilon/2)` | passes 80% of the time |
| `(2^ell+epsilon) - 1` | fails |
| `2^ell+epsilon` | fails |

I decided to add a test that `2^ell` always passes, and `2^ell+epsilon` always fails, since I think those two things both happen with high probability.

Reply:
Interesting! I'll need to dig into the paper a bit to understand what the claim _really_ is, because Figure 14 makes it sound like there's a guarantee that `k \in 2^ell` at the top, but the range check says "The proof guarantees that k \in 2^{ell + epsilon}`, which it sounds like it doesn't necessarily...

Reply:
Yeah, I took a quick look at the proof of soundness but I don't know all the notation they're using, will have to look more carefully at some point.